src/webgpu/api/validation/vertex_access.spec.ts

```diff
@@ -47,45 +59,545 @@ TODO: make sure this isn't already covered somewhere else, review, organize, and
 >         - bound vertex buffer **ranges** are {exact size, just under exact size} needed for draws with:
->             - vertexCount largeish
->             - firstVertex {=, >} 0
->             - instanceCount largeish
->             - firstInstance {=, >} 0
->         - include VBs with both step modes
->     - x= {draw, drawIndirect}
-> - indexed draws:
->     - vertex access out of bounds (make sure this doesn't overlap with robust access)
->         - bound vertex buffer **ranges** are {exact size, just under exact size} needed for draws with:
->             - a vertex index in the buffer is largeish
->             - baseVertex {=, >} 0
->             - instanceCount largeish
->             - firstInstance {=, >} 0
->         - include VBs with both step modes
->     - x= {drawIndirect, drawIndexedIndirect}
+>                 - a vertex index in the buffer is largeish
+>                 - baseVertex {=, >} 0
+>                 - instanceCount largeish
+>                 - firstInstance {=, >} 0
+>             - include VBs with both step modes
+>         - x= {render pass, render bundle}
+> - In queue.submit():
+>     - Every GPUBuffer referenced in any element of commandBuffers is in the "unmapped" buffer state.
 `;
 
 import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { GPUConst } from '../../constants.js';
 
 import { ValidationTest } from './validation_test.js';
 
-export const g = makeTestGroup(ValidationTest);
+// A map for buffer format informations. compatibleType indicate the general class of a format, and
+// can further map to compatible and incompatible WGSL types. I.e., for "float", "f32" and "vec*<f32>"
+// are compatible WGSL types, while "i32" and others are incompatible WGSL types.
+const typeInfoMap: {
+  [k: string]: { wgslType: string; sizeInBytes: number; compatibleType: string };
+} = {
+  float32: {
+    wgslType: 'f32',
+    sizeInBytes: 4,
+    compatibleType: 'float',
+  },
+  float32x2: {
+    wgslType: 'vec2<f32>',
+    sizeInBytes: 8,
+    compatibleType: 'float',
+  },
+  float32x3: {
+    wgslType: 'vec3<f32>',
+    sizeInBytes: 12,
+    compatibleType: 'float',
+  },
+  float32x4: {
+    wgslType: 'vec4<f32>',
+    sizeInBytes: 16,
+    compatibleType: 'float',
+  },
+  uint8x2: {
+    wgslType: 'vec2<u32>',
+    sizeInBytes: 2,
+    compatibleType: 'uint',
+  },
+  uint8x4: {
+    wgslType: 'vec4<u32>',
+    sizeInBytes: 4,
+    compatibleType: 'uint',
+  },
+  unorm8x2: {
+    wgslType: 'vec2<f32>',
+    sizeInBytes: 2,
+    compatibleType: 'float',
+  },
+  unorm8x4: {
+    wgslType: 'vec4<f32>',
+    sizeInBytes: 4,
+    compatibleType: 'float',
+  },
+};
+
+// Class that indicate how to make input attribute for vertex shader module
+interface VertexBufferDescriptorForWGSLShader {
+  shaderLocation: GPUIndex32;
+  wgslType: string;
+}
+
+// Class that indicate how to call a draw function
+class DrawCall {
+  test: ValidationTest;
```

Comment:
Yes, it is unused yet, but when it comes to drawIndirect and drawIndexedIndirect, it may use test to create the corresponding indirect buffer. I removed the code of DrawCall in the new commit to keep it simple.

Reply:
OK, no problem. OK to have dead code if you will use it soon.